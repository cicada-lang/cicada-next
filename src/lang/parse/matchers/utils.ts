import * as Exps from "../../exp"

export function piBindingtoFnBindingFrom(binding: Exps.PiBinding): Exps.FnBinding {
  switch (binding.kind) {
    case "PiBindingNameless": {
      return Exps.FnBindingName("_")
    }

    case "PiBindingNamed": {
      return Exps.FnBindingName(binding.name)
    }

    case "PiBindingImplicit": {
      return Exps.FnBindingImplicit(binding.name)
    }
  }
}
