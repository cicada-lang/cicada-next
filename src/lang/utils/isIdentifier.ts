const identifierRegExp = new RegExp("^([-_\\p{Letter}][-_\\p{Letter}0-9]*)$", "u")

export function isIdentifier(text: string): boolean {
  return identifierRegExp.test(text)
}
