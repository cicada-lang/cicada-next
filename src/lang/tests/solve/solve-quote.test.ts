import { expect, test } from "vitest"
import { runCode } from "../utils"

test("solve Quote -- data", async () => {
  const output = await runCode(`

solve () {
  unify "a" = "a"
  unify "b" = "b"
  unify "c" = "c"
}

`)

  expect(output).toMatchInlineSnapshot('"{  }"')
})

test("solve Quote -- bindings", async () => {
  const output = await runCode(`

solve (a: String, b: String, c: String) {
  unify a = "a"
  unify b = "b"
  unify c = "c"
}

`)

  expect(output).toMatchInlineSnapshot('"{ a: \\"a\\", b: \\"b\\", c: \\"c\\" }"')
})

test("solve Quote -- walk", async () => {
  const output = await runCode(`

solve (a: String, b: String, c: String) {
  unify a = b
  unify b = c
  unify a = "a"
}

`)

  expect(output).toMatchInlineSnapshot('"{ a: \\"a\\", b: \\"a\\", c: \\"a\\" }"')
})
