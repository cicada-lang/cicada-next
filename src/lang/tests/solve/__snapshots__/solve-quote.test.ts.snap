// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`solve Quote -- bindings 1`] = `
"{
  a: "a",
  b: "b",
  c: "c"
}"
`;

exports[`solve Quote -- data 1`] = `"{}"`;

exports[`solve Quote -- walk 1`] = `
"{
  a: "a",
  b: "a",
  c: "a"
}"
`;
